util/hookable/hookable.cpp

```diff
@@ -213,39 +216,56 @@ J9HookDispatch(struct J9HookInterface **hookInterface, uintptr_t taggedEventNum,
 			/* now read the id again to make sure that nothing has changed */
 			VM_AtomicSupport::readBarrier();
 			if (record->id == id) {
-				uint64_t timeDelta = 0;
+				uint64_t startTime = 0;
+				uint64_t count = 0;
+				if (NULL != eventDump) {
+					count = VM_AtomicSupport::addU64(&eventDump->count,1);
+					sampling = (1 >= samplingInterval) || ((100 >= samplingInterval) && (0 == (count ^ samplingInterval)));
+				} else {
+					sampling =  false;
+				}
 				OMRPORT_ACCESS_FROM_OMRPORT(commonInterface->portLib);
-				uint64_t startTime = omrtime_current_time_millis();
+				if (sampling) {
+					startTime = omrtime_current_time_millis();
+				}
+
 				function(hookInterface, eventNum, eventData, userData);
-				timeDelta = omrtime_current_time_millis() - startTime;
-
-				OMREventInfo4Dump *eventDump = J9HOOK_DUMPINFO(commonInterface, eventNum);
-
-				/* record hook info for dump if elapse time is longer than 1 millisecond */
-				if ((NULL != eventDump) && (0 != timeDelta)) {
-					eventDump->lastHook.callsite = record->callsite;
-					eventDump->lastHook.func_ptr = (void *)record->function;
-					eventDump->lastHook.startTime = startTime;
-					eventDump->lastHook.duration = timeDelta;
-					if (eventDump->longestHook.duration < eventDump->lastHook.duration) {
-						eventDump->longestHook.callsite = eventDump->lastHook.callsite;
-						eventDump->longestHook.startTime = eventDump->lastHook.startTime;
-						eventDump->longestHook.func_ptr = eventDump->lastHook.func_ptr;
-						eventDump->longestHook.duration = eventDump->lastHook.duration;
+
+				if (sampling) {
+					uint64_t timeDelta = omrtime_current_time_millis() - startTime;
+					uint64_t lastStart = 0;
+
+					/* here is not thread safe, keep eventDump data consistency from race condition */
+					do {
```

Comment:
I do not feel like this is the right way to update this structure.  Some fields are updated outside of the atomic and there are two atomic operations. Let me think a bit about this to see a bit more.